src/grisp_connect_connection.erl

```diff
@@ -0,0 +1,741 @@
+
+% API Functions
+-export([start_link/2]).
+-export([request/3]).
+-export([post/4]).
+-export([notify/3]).
+-export([reply/3]).
+-export([error/5]).
+-export([disconnect/1]).
+
+% Behaviour gen_server Callbacks
+-export([init/1]).
+-export([handle_call/3]).
+-export([handle_cast/2]).
+-export([handle_info/2]).
+-export([terminate/2]).
+
+
+%--- Types ---------------------------------------------------------------------
+
+-record(batch, {
+    bref :: reference(),
+    refcount :: pos_integer(),
+    responses :: list()
+}).
+
+-record(inbound_req, {
+    method :: method(),
+    id :: binary() | integer(),
+    bref :: undefined | reference() % Set if part of a batch
+}).
+
+-record(outbound_req, {
+    method :: method(),
+    id :: binary() | integer(),
+    tref :: undefined | reference(),
+    from :: undefined | gen_server:from(),
+    ctx :: undefined | term()
+}).
+
+-record(state, {
+    handler :: pid(),
+    uri :: iodata(),
+    domain :: binary(),
+    port :: inet:port_number(),
+    path :: binary(),
+    ping_timeout :: infinity | pos_integer(),
+    request_timeout :: infinity | pos_integer(),
+    batches = #{} :: #{reference() => #batch{}},
+    inbound = #{} :: #{binary() | integer() => #inbound_req{}},
+    outbound = #{} :: #{binary() | integer() => #outbound_req{}},
+    gun_pid :: undefined | pid(),
+    gun_ref :: undefined | reference(),
+    ws_stream :: undefined | gun:stream_ref(),
+    connected = false :: boolean(),
+    ping_tref :: undefined | reference()
+}).
+
+-type error_mapping() :: {atom(), integer(), binary()}.
+-type method() :: atom() | binary() | [atom() | binary()].
+-type start_options() :: #{
+    domain := atom() | string() | binary(),
+    port := inet:port_number(),
+    %FIXME: dialyzer do no like ssl:tls_client_option(), maybe some erlang version issue
+    transport := tcp | tls | {tls, TlsOpts :: list()},
+    path := atom() | string() | binary(),
+    errors => [error_mapping()],
+    ping_timeout => infinity | pos_integer(),
+    request_timeout => infinity | pos_integer()
+}.
+
+% Type specfication of the messages that are sent to the handler:
+-type handler_messages() ::
+    {conn, pid(), connected}
+  | {conn, pid(), {request, method(), Params :: map() | list(), ReqRef :: binary() | integer()}}
+  | {conn, pid(), {notification, method(), Params :: map() | list()}}
+  | {conn, pid(), {response, Result :: term(), Ctx :: term()}}
+  | {conn, pid(), {remote_error, Code :: integer() | atom(), Message :: undefined | binary(), Data :: term()}}
+  | {conn, pid(), {remote_error, Code :: integer() | atom(), Message :: undefined | binary(), Data :: term(), Ctx :: term()}}
+  | {conn, pid(), {local_error, Reason:: atom(), Ctx :: term()}}.
+
+-export_type([error_mapping/0, method/0, handler_messages/0]).
+
+
+%--- Macros --------------------------------------------------------------------
+
+-define(ENABLE_TRACE, false).
+-if(?ENABLE_TRACE =:= true).
```

Comment:
So is it possible to step over written definitions?

Reply:
@ziopio Not sure what you mean by that...

Reply:
You define `ENABLE_TRACE = false` and then use it right away to check if it is true. Without context this looks like the if section will never apply.